src/Microsoft.AspNetCore.Authentication/AuthenticationHandler.cs

```diff
@@ -56,193 +56,104 @@ protected string CurrentUri
-            UrlEncoder = encoder;
-
-            RegisterAuthenticationHandler();
-
-            Response.OnStarting(OnStartingCallback, this);
 
-            if (ShouldHandleScheme(AuthenticationManager.AutomaticScheme, Options.AutomaticAuthenticate))
+            Options = OptionsSnapshot.Get(Scheme.Name) ?? new TOptions();
+            if (!Options.Initialized)
             {
-                InitializeResult = await HandleAuthenticateOnceAsync();
-                if (InitializeResult?.Skipped == true || InitializeResult?.Handled == true)
-                {
-                    return;
-                }
-
-                if (InitializeResult?.Failure != null)
+                lock (Options.InitializeLock)
                 {
-                    Logger.AuthenticationSchemeNotAuthenticatedWithFailure(Options.AuthenticationScheme, InitializeResult.Failure.Message);
-                }
-
-                var ticket = InitializeResult?.Ticket;
-                if (ticket?.Principal != null)
-                {
-                    Context.User = SecurityHelper.MergeUserPrincipal(Context.User, ticket.Principal);
-                    Logger.UserPrinicpalMerged(Options.AuthenticationScheme);
+                    if (!Options.Initialized)
+                    {
+                        InitializeOptions();
+                        Options.Initialized = true;
+                    }
                 }
             }
-        }
-
-        protected string BuildRedirectUri(string targetPath)
-        {
-            return Request.Scheme + "://" + Request.Host + OriginalPathBase + targetPath;
-        }
 
-        private static async Task OnStartingCallback(object state)
-        {
-            var handler = (AuthenticationHandler<TOptions>)state;
-            await handler.FinishResponseOnce();
-        }
+            // REVIEW: should we validate options only once inside of the lock?
+            Options.Validate();
 
-        private async Task FinishResponseOnce()
-        {
-            if (!_finishCalled)
-            {
-                _finishCalled = true;
-                await FinishResponseAsync();
-                await HandleAutomaticChallengeIfNeeded();
-            }
+            return InitializeEventsAsync();
         }
 
         /// <summary>
-        /// Hook that is called when the response about to be sent
+        /// Initializes the events object, called once per request by <see cref="InitializeAsync(AuthenticationScheme, HttpContext)"/>.
         /// </summary>
-        /// <returns></returns>
-        protected virtual Task FinishResponseAsync()
+        protected virtual async Task InitializeEventsAsync()
         {
-            return TaskCache.CompletedTask;
-        }
-
-        private async Task HandleAutomaticChallengeIfNeeded()
-        {
-            if (!ChallengeCalled && Options.AutomaticChallenge && Response.StatusCode == 401)
+            Events = Options.Events;
+            if (Options.EventsType != null)
```

Comment:
Yeah so Events is always not null in general, as the default is an empty events object, if we added that check, they'd have to explicitly null out events and set eventstype.